quinn-proto/src/connection/assembler.rs

```diff
@@ -144,20 +127,71 @@ impl Assembler {
         if bytes.is_empty() {
             return;
         }
-
-        self.data.push(Buffer { offset, bytes });
-        // Why 32: on the one hand, we want to defragment rarely, ideally never
+        let buffer = Buffer::new(offset, bytes, allocation_size);
```

Comment:
What is the reason for not calling `buffer.prepare_defragment` here, or even executing the logic as part of `Buffer::new()`?

Reply:
Thinking more about it, is it to capture the situation where buffers are fully filling the allocation and are therefore not oversized, but where the overlap between buffers requires the defragmentation?

Reply:
Yes, if we would mark a `Buffer` immediately as defragmented, we cannot get rid of it in `defragment`, if e.g. it contains only one byte after discarding duplicate data.